qiita_db/sql_connection.py

```diff
@@ -460,176 +397,202 @@ def execute_fetchall(self, sql, sql_args=None):
 
         return result
 
-    def _check_queue_exists(self, queue_name):
-        """Checks if queue `queue_name` exists in the handler
 
-        Parameters
-        ----------
-        queue_name : str
-            The name of the queue
+class Transaction(object):
+    """Encapsulates a DB transaction
 
-        Returns
-        -------
-        bool
-            True if queue `queue_name` exist in the handler. False otherwise.
-        """
-        return queue_name in self.queues
+    A transaction is defined by a series of consecutive queries that need to
+    be applied to the database as a single block.
+
+    Parameters
+    ----------
+    name : str
+        Name of the transaction.
+    """
 
-    def create_queue(self, queue_name):
-        """Add a new queue to the connection
+    _regex = re.compile("^{(\d+):(\d+):(\d+)}$")
 
-        Parameters
-        ----------
-        queue_name : str
-            Name of the new queue
+    def __init__(self, name):
+        # The name is useful for debugging, since we can identify the
+        # failed queue in errors
+        self._name = name
+        self._queries = []
+        self._results = []
+        self._index = 0
+        self._conn_handler = SQLConnectionHandler()
+
+    def _raise_execution_error(self, sql, sql_args, error):
+        """Rollbacks the current transaction and raises a useful error
+
+        The error message contains the name of the transaction, the failed
+        query, the arguments of the failed query and the error generated.
 
         Raises
         ------
-        KeyError
-            Queue name already exists
+        ValueError
         """
-        if self._check_queue_exists(queue_name):
-            raise KeyError("Queue %s already exists" % queue_name)
+        self.rollback()
+        raise ValueError(
+            "Error running SQL query in transaction %s:\n"
+            "Query: %s\nArguments: %s\nError: %s\n"
+            % (self._name, sql, str(sql_args), str(error)))
 
-        self.queues[queue_name] = []
+    def _replace_placeholders(self, sql, sql_args):
+        """Replaces the placeholder in `sql_args` with the actual value
 
-    def list_queues(self):
-        """Returns list of all queue names currently in handler
+        Parameters
+        ----------
+        sql : str
+            The SQL query
+        sql_args : list or None
```

Comment:
Uhm, good point. At this point, the parameter `sql_args` comes from the user, but I think one of the first checks that I can do when that parameters gets added to the transaction is to check that if it is None, I can put an empty list. psycopg2 does not file passing an empty list or None, so it will simplify code. Good suggestion! Thanks,